### ISSUE TITLE
Possibility to configure Dokka 1.4  by other Gradle plugin

### ISSUE THREAD

**Issue by tprochazka** [Kotlin/dokka#1463]
I have own Kotlin based Gradle plugin, which setup everything for all my projects (publishing, signing, documentation, etc.)
Including Dokka.
Now I'm migration from 0.10.1 to 1.4.0.

And attempts to generate doc ends with this:
```
> Task :sample-library:dokkaHtml
Dokka 1.4.* is an alpha project
Initializing plugins
Creating documentation models
Exiting Generation: Nothing to document
```

I found that problem is in missing `context.configuration.sourceSets` in method `createDocumentationModels`.
So it looks that for some reason automatic sourceRoots detection doesn't works in this case.

Here is test project [dokka-test.zip](https://github.com/Kotlin/dokka/files/5211676/dokka-test.zip)

I found that the key difference is the place when I put your plugin to classpath. If I will do that inside of project gradle script by

```
buildscript {
    repositories {
        jcenter()
    }
    dependencies {
        classpath("org.jetbrains.dokka:dokka-gradle-plugin:1.4.0")
    }
}
```

It works correctly, but when I put it to the `buildSrc/build.gradle.kts`  (to simulate plugin behavior)
```
dependencies {
    implementation("org.jetbrains.dokka:dokka-gradle-plugin:1.4.0")  
}
```

It still works, dokka task is running, but it ends with `Exiting Generation: Nothing to document`
The automatic source roots setup is not working.

**Comment by tprochazka** [comment]
Can you navigate me to the location in the source code where automatic source roots detection happens please?

**Comment by tprochazka** [comment]
Please I can debug and try to fix (found solution) it myself. But I need a little bit of help to understand how it should work and where found the loggic responsible for root source setup.

**Comment by kamildoleglo** [comment]
Sorry for the late reply. I suppose that the gradle lifecycle may be an issue here. Nevertheless, the source roots detection is done in the `DokkaTask`'s `val dokkaSourceSets` using the `configureWithKotlinSourceSet` method and extracts the source roots, classpath and so on from the Kotlin Plugin in the project

**Comment by tprochazka** [comment]
Thanks for your response! I was so busy until now. I just tried it and the first issue is that `project.kotlinOrNull` inside of `DokkaTask` simply returns `null`, so `configureWithKotlinSource` is newer called then. It looks that it happens directly in the configuration time before Kotlin plugin is actually applied. I'm not sure how you ensure that Dokka wait to Kotlin plugin initialization. I'm using this way in my own plugin

project.pluginManager.withPlugin("org.jetbrains.kotlin.android") {
  ... stuff which I need to do when Kotlin is available ...
}

In this way, Gradle notifies me, just at the moment when the plugin is applied.
Maybe you are doing something else, that doesn't work if I apply the Dokka plugin manually inside of my own plugin.

I will try to postpone the Dokka plugin applying until Kotlin is applied. But it is probably not a good idea, because I want to have Dokka working also on the modules, that are still not converted to Kotlin and using just Java.

**Comment by tprochazka** [comment]
Interesting. Even when I do this

        project.pluginManager.withPlugin("org.jetbrains.kotlin.android") {
            project.run {
                pluginManager.apply("org.jetbrains.dokka")

                tasks.withType(DokkaTask::class.java).configureEach {
                    dependencies {
                        plugins("org.jetbrains.dokka:kotlin-as-java-plugin:1.4.0")
                    }

                    outputDirectory.set(buildDir.resolve("docs/dokka"))
                    dokkaSourceSets.configureEach {
                        jdkVersion.set(8)
                        platform.set(org.jetbrains.dokka.Platform.jvm)
                    }
                }
            }
        }

`project.kotlinOrNull` still return `null`

But inside of `project.extensions` collection is already a `kotlin` item, as you can see here in debugger:
![image](https://user-images.githubusercontent.com/172094/100809997-bfc2ea80-3437-11eb-9583-bd803addd9e5.png)

**Comment by tprochazka** [comment]
It is probably the same problem that I found during the development of my plugin.
If I do this in the project

plugins {
    id("my.plugin") version "X.Y.Z"
    id("org.jetbrains.kotlin.jvm") version "1.4.0"
}

Gradle will use a different class loader for both plugins and then class `KotlinProjectExtension::class.java` is invisible inside of your method in Dokka plugin.

Question is why it works in this case
plugins {
    kotlin("jvm") version "1.4.10"
    id("org.jetbrains.dokka") version "1.4.10.2"
}

Because both plugins are `org.jetbrains` Gradle will use the same classloader?

**Comment by tprochazka** [comment]
When I put Dokka and Kotlin on the classpath of my plugin and use plugins {} declaration without version, it works.
But nobody is able to explain to me, why Dokka a and Kotlin itself specified by version, works together. @kamildoleglo

**Comment by kamildoleglo** [comment]
Not a clue actually. I'll try to do my research tomorrow and I'll get back to you

**Comment by tprochazka** [comment]
Thanks. I tried to ask also on Gradle slack
https://gradle-community.slack.com/archives/CA745PZHN/p1606867658232400

**Comment by fondesa** [comment]
The exact same issue is happening to us.
@tprochazka In the meanwhile, have you found a way to workaround it?

**Comment by tprochazka** [comment]
@fondesa No, the only solution so far is don't use plugins definition with version, just put all plugin on classpath directly in the root build script, in the old way. :-( I'm trying to solve it also directly in the Gradle Slack and it looks that they have no solution for it. It is very sad :-(

**Comment by tprochazka** [comment]
I got information from the Gradle developer, that this issue happens only when my plugin is defined by `buildSrc`. To replace it by `includeBuild` or deploy a plugin to some move repo and use it from there should be fine.  But I did not test it yet.

**Comment by elect86** [comment]
> I got information from the Gradle developer, that this issue happens only when my plugin is defined by `buildSrc`. To replace it by `includeBuild` or deploy a plugin to some move repo and use it from there should be fine. But I did not test it yet.

Sorry, I'm not sure I got the workarounds right: use composite builds or move code in a dedicated plugin?

**Comment by tprochazka** [comment]
> use composite builds or move code in a dedicated plugin?
Both ways should work, the only problem is with `buildSrc`

**Comment by fondesa** [comment]
> I got information from the Gradle developer, that this issue happens only when my plugin is defined by buildSrc. To replace it by includeBuild or deploy a plugin to some move repo and use it from there should be fine. But I did not test it yet.

I confirm that moving the plugin to a composite build indeed fixes the problem.

**Comment by swankjesse** [comment]
Please reopen this issue and implement a fix? It’s quite surprising to have a plugin that stops working when you install it differently.

**Comment by tprochazka** [comment]
It is not this plugin Issue, it is a general Gradle issue when you using deprecated `buildSrc`.
So it is not possible to fix it on the plugin side.

**Comment by swankjesse** [comment]
> deprecated buildSrc

Is it? Mind linking to the general Gradle issue? I was hoping for a general way to configure my projects and this is the only thing blocking that.

**Comment by tprochazka** [comment]
> > deprecated buildSrc
> 
> Is it? Mind linking to the general Gradle issue? I was hoping for a general way to configure my projects and this is the only thing blocking that.

Yes. Instead of buildSrc, you can use composite/included build and it works with that.

**Comment by syphyr** [comment]
I am also trying to work around this bug with dokka 2.1.0. I have tried adding dokka to the `classpath` in build.gradle, instead of adding to the plugins, but it does not help. The subproject contains build.gradle.kts and I am unable to load kotlin-android before dokka without causing conflicts with the project. Does anyone have any suggestions?

Edit: I fixed my issue by using kotlin-android instead of kotlin.jvm.
